io/mpi/src/TMPIFile.cxx

```diff
@@ -0,0 +1,592 @@
+   return fClientsContact.CountBits() > cut || fNClientsContact > 2 * cut;
+}
+
+void TMPIFile::R__MigrateKey(TDirectory *destination, TDirectory *source)
+{
+   if (destination == 0 || source == 0)
+      return;
+   TIter nextkey(source->GetListOfKeys());
+   TKey *key;
+   while ((key = (TKey *)nextkey())) {
+      TClass *cl = TClass::GetClass(key->GetClassName());
+      if (cl->InheritsFrom(TDirectory::Class())) {
+         TDirectory *source_subdir = (TDirectory *)source->GetList()->FindObject(key->GetName());
+         if (!source_subdir) {
+            source_subdir = (TDirectory *)key->ReadObj();
+         }
+         TDirectory *destination_subdir = destination->GetDirectory(key->GetName());
+         if (!destination_subdir) {
+            destination_subdir = destination->mkdir(key->GetName());
+         }
+         TMPIFile::R__MigrateKey(destination, source);
+      } else {
+         TKey *oldkey = destination->GetKey(key->GetName());
+         if (oldkey) {
+            oldkey->Delete();
+            delete oldkey;
+         }
+         TKey *newkey = new TKey(destination, *key, 0 /* pidoffset */);
+         destination->GetFile()->SumBuffer(newkey->GetObjlen());
+         newkey->WriteFile(0);
+         if (destination->GetFile()->TestBit(TFile::kWriteError)) {
+            return;
+         }
+      }
+   }
+   destination->SaveSelf();
+}
+
+void TMPIFile::R__DeleteObject(TDirectory *dir, Bool_t withReset)
+{
+   if (dir == 0)
+      return;
+
+   TIter nextkey(dir->GetListOfKeys());
+   TKey *key;
+   while ((key = (TKey *)nextkey())) {
+      TClass *cl = TClass::GetClass(key->GetClassName());
+      if (cl->InheritsFrom(TDirectory::Class())) {
+         TDirectory *subdir = (TDirectory *)dir->GetList()->FindObject(key->GetName());
+         if (!subdir) {
+            subdir = (TDirectory *)key->ReadObj();
+         }
+         TMPIFile::R__DeleteObject(subdir, withReset);
+      } else {
+         Bool_t todelete = kFALSE;
+         if (withReset) {
+            todelete = (0 != cl->GetResetAfterMerge());
+         } else {
+            todelete = (0 == cl->GetResetAfterMerge());
+         }
+         if (todelete) {
+            key->Delete();
+            dir->GetListOfKeys()->Remove(key);
+            delete key;
+         }
+      }
+   }
+}
+
+////////////////////////////////////////////////////////////////////////////////
+/// return True if this is the Collector rank, otherwise False
+
+Bool_t TMPIFile::IsCollector()
+{
+   if (this->fMPILocalRank) {
+      return kFALSE;
+   }
+   return kTRUE;
+}
+
+////////////////////////////////////////////////////////////////////////////////
+/// Called by the Workers only: Copies the current content in memory and
+/// sends it asynchronously to the Collector for merging and writing to disk.
+///
+
+void TMPIFile::CreateBufferAndSend()
+{
+   if (this->IsCollector()) {
+      SysError("CreateBufferAndSend", " should not be called by a collector");
+      exit(1);
```

Comment:
Please use `Error` instead of `SysError` and don't call `exit(1)`. We probably don't want to quit the interpreter if people are using that.